transport/ice/IceSession.cpp

```diff
@@ -472,11 +493,252 @@ bool IceSession::isResponseAuthentic(const void* data, size_t len) const
+    {
+        candidatePair->accept();
+        sendResponse(localEndpoint, remotePort, 0, msg, now);
+        candidatePair->receptionTimestamp = now;
+        candidatePair->send(now);
+    }
+
+    if (_state == State::CONNECTING)
+    {
+        sortCheckList();
+    }
+
+    return candidatePair != nullptr;
+}
+
+bool IceSession::processValidStunUdpRequest(IceEndpoint* localEndpoint,
+    const transport::SocketAddress& remotePort,
+    const StunMessage& msg,
+    int remoteCandidatePriority,
+    uint64_t now)
+{
+    assert(localEndpoint->getTransportType() == TransportType::UDP);
+
+    if (!(_state == State::CONNECTING || _state == State::CONNECTED))
+    {
+        return false;
+    }
+
+    auto* pair = findCandidatePair(localEndpoint, remotePort);
+    if (pair)
+    {
+        // Only send response after accept! Otherwise we could drop early media
+        pair->accept();
+        sendResponse(localEndpoint, remotePort, 0, msg, now);
+        return true;
+    }
+
+    const auto* remoteCandidate = findCandidateWithUdpAddress(_remoteCandidates, remotePort);
+    if (remoteCandidate)
+    {
+        // if it is not a new candidate but we don't have a candidate that could mean we are receiving
+        // a request from a pair we have discarded previously.
+        // I don't see why this should happen but if we are receiving a probe from it, perhaps there is no reason to be
+        // discarded. Let's add it again and log a warn
+        const auto* localEndpointInfo = findEndpointInfo(localEndpoint);
+        assert(localEndpointInfo);
+        if (!localEndpointInfo)
+        {
+            // We can't accept endpoint that are not attached. This should be an error as it is not supposed to
+            // happen
+            // This perhaps should send an error instead of just ignore
+            return false;
+        }
+
+        auto* candidatePair = addProbeForRemoteCandidate(*localEndpointInfo, *remoteCandidate);
+        if (candidatePair)
+        {
+            candidatePair->accept();
+            candidatePair->nextTransmission = now;
+            candidatePair->receptionTimestamp = now;
+
+            // Only send response after accept! Otherwise we could drop early media
+            sendResponse(localEndpoint, remotePort, 0, msg, now);
+            sortCheckList();
+            processTimeout(now);
+            return true;
+        }
+    }
+
+    // If we are here, it is a new candidate
+
+    if (_state == State::CONNECTED || _remoteCandidates.size() >= _config.maxCandidateCount)
+    {
+        // When we see a new candidate and the ICE is CONNECTED, this can be an early sign that network state
+        // has changed on client. Therefore, we will try to remove some candidates that has not being received
+        // probes nor media for a while to ensure we can had new candidates. The removing will be more
+        // aggressive if the _remoteCandidates is full
+        removeNonReadableRemoteCandidates(now);
```

Comment:
If the candidate list is full or the state is connected (perhaps is an early indication that something has changed on client with a new candidate being available). We are going to try to reduce the candidate list